File: ride-service/src/main/java/com/example/ride_service/exception/GlobalExceptionHandler.java
package com.example.ride_service.exception;

import com.example.ride_service.exception.models.ErrorResponse;
import com.example.ride_service.exception.models.ValidationError;
import com.example.ride_service.exception.models.ValidationErrorResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.stream.Collectors;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<ErrorResponse> handleEntityNotFound(NotFoundException ex) {
        ErrorResponse error = new ErrorResponse(
                "NOT_FOUND",
                ex.getMessage()
        );
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ValidationErrorResponse> handleValidationErrors(MethodArgumentNotValidException ex) {
        ValidationErrorResponse response = new ValidationErrorResponse(
                "VALIDATION_FAILED",
                "One or more fields are invalid",
                ex.getBindingResult()
                        .getFieldErrors()
                        .stream()
                        .map(error -> ValidationError.builder()
                                .field(error.getField())
                                .message(error.getDefaultMessage()).build()
                        )
                        .collect(Collectors.toList())
        );
        return ResponseEntity.badRequest().body(response);
    }

    @ExceptionHandler(EstimateExpiredException.class)
    public ResponseEntity<ErrorResponse> handleEstimateExpired(EstimateExpiredException ex) {
        ErrorResponse error = new ErrorResponse(
                "ESTIMATE_EXPIRED",
                ex.getMessage()
        );
        return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(error);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleAll(Exception ex) {
        log.error(ex.getMessage(), ex);
        return ResponseEntity
                .status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ErrorResponse("INTERNAL_SERVER_ERROR",
                        "Unexpected error occurred")
                );
    }
}

File: driver-service/src/main/kotlin/com/example/driver_service/exception/GlobalExceptionHandler.kt
package com.example.driver_service.exception

import com.example.driver_service.exception.models.ErrorResponse
import com.example.driver_service.exception.models.ValidationError
import com.example.driver_service.exception.models.ValidationErrorResponse
import mu.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {
    companion object {
        private val log = KotlinLogging.logger {}
    }

    @ExceptionHandler(EntityNotFoundException::class)
    fun handleEntityNotFoundException(ex: EntityNotFoundException):
            ResponseEntity<ErrorResponse> {
        return ResponseEntity(
            ErrorResponse("NOT_FOUND", ex.message!!),
            HttpStatus.NOT_FOUND
        )
    }

    @ExceptionHandler(InvalidCredentialsException::class)
    fun handleInvalidCredentialsException(ex: InvalidCredentialsException):
            ResponseEntity<ErrorResponse> {
        return ResponseEntity(
            ErrorResponse("UNAUTHORIZED", ex.message!!),
            HttpStatus.UNAUTHORIZED
        )
    }

    @ExceptionHandler(CarLimitExceededException::class)
    fun handleCarLimitExceededException(ex: CarLimitExceededException):
            ResponseEntity<ErrorResponse> {
        return ResponseEntity(
            ErrorResponse("CAR_LIMIT_EXCEEDED", ex.message!!),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValidException(ex: MethodArgumentNotValidException):
            ResponseEntity<ValidationErrorResponse> {
        return ResponseEntity(
            ValidationErrorResponse(
                "VALIDATION_FAILED",
                "One or more fields are invalid",
                ex.bindingResult
                    .fieldErrors
                    .stream()
                    .map { error
                        ->
                        ValidationError(error.field, error.defaultMessage!!)
                    }
                    .toList()
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(ResourceAlreadyExistsException::class)
    fun handleResourceAlreadyExistsException(ex: ResourceAlreadyExistsException):
            ResponseEntity<ErrorResponse> {
        return ResponseEntity(
            ErrorResponse("CONFLICT", ex.message!!),
            HttpStatus.CONFLICT
        )
    }

    @ExceptionHandler(Exception::class)
    fun handleAllExceptions(ex: Exception): ResponseEntity<ErrorResponse> {
        log.error { "SERVER ERROR: ${ex.message}" }
        return ResponseEntity(
            ErrorResponse("SERVER_ERROR", "Unknown error"),
            HttpStatus.INTERNAL_SERVER_ERROR
        )
    }
}